### PR TITLE
feat: increase test coverage on reconciler

### PR DESCRIPTION
**Pull request by leoparente**
- Add `miniredis` to emulate redis connection and test public methods accordingly
- Miniredis does not support RedisJSON, so internal package test are required
- Add internal tests to properly cover `ingestion_processor.go`
- interface `redis.Client` used methods with `RedisClient` to allow proper mocking
- replace `*netboxdiodeplugin.Client` with the interface `netboxdiodeplugin.NetBoxAPI`

### CODE REVIEW

**Comment by leoparente** [comment]
9% increase:
image

File: diode-server/go.mod [hunk 1]
```diff
@@ -20,6 +20,12 @@ require (
 )
 
 require (
+	github.com/alicebob/gopher-json v0.0.0-20200520072559-a9ecdc9d1d3a // indirect
+	github.com/yuin/gopher-lua v1.1.1 // indirect
+)
+
+require (
```
Comment (mfiedorowicz) [inline diode-server/go.mod]: Could we please keep these within one `require ()` block for indirects?

File: diode-server/go.mod [hunk 2]
```diff
@@ -20,6 +20,12 @@ require (
...
+	github.com/alicebob/miniredis/v2 v2.33.0
```
Comment (mfiedorowicz) [inline diode-server/go.mod]: This one, as not indirect, should be in the `require ()` block for non-indirects.

File: diode-server/reconciler/ingestion_processor_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+	"testing"
+
+	"github.com/netboxlabs/diode/diode-server/gen/diode/v1/diodepb"
+	"github.com/netboxlabs/diode/diode-server/netbox"
+	"github.com/netboxlabs/diode/diode-server/netboxdiodeplugin"
+	"github.com/netboxlabs/diode/diode-server/reconciler/changeset"
+	"github.com/redis/go-redis/v9"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/proto"
```
Comment (mfiedorowicz) [inline diode-server/reconciler/ingestion_processor_internal_test.go]: Could we keep imports organised in the groups?
```
stdlib

external imports

internal imports
```

File: diode-server/reconciler/ingestion_processor_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+
+func (m *MockRedisClient) XDel(ctx context.Context, stream string, ids ...string) *redis.IntCmd {
+	args := m.Called(ctx, stream, ids)
+	return args.Get(0).(*redis.IntCmd)
+}
+
+func (m *MockRedisClient) Do(ctx context.Context, args ...interface{}) *redis.Cmd {
+	callArgs := m.Called(ctx, args)
+	return callArgs.Get(0).(*redis.Cmd)
+}
```
Comment (mfiedorowicz) [inline diode-server/reconciler/ingestion_processor_internal_test.go]: For `netboxdiodeplugin.NetBoxAPI` we already have generated in https://github.com/netboxlabs/diode/blob/develop/diode-server/netboxdiodeplugin/mocks/netboxapi.go so not needed here.

In general, these mocks should be generated for using [Mockery](https://vektra.github.io/mockery/latest/), apologies it's not documented (yet)

- installation
```shell
go install github.com/vektra/mockery/v2@v2.43.2
```

- update https://github.com/netboxlabs/diode/blob/develop/diode-server/.mockery.yaml by appending following to `packages`:
```
  github.com/netboxlabs/diode/diode-server/reconciler:
    config:
      all: true
      dir: "{{.InterfaceDir}}/mocks"
      outpkg: mocks
```
- generate mocks:
```shell
mockery
```

- update tests to import and use mocks

File: diode-server/reconciler/ingestion_processor.go [hunk 5]
```diff
@@ -44,14 +44,25 @@ const (
...
 type IngestionProcessor struct {
 	config            Config
 	logger            *slog.Logger
 	hostname          string
-	redisClient       *redis.Client
-	redisStreamClient *redis.Client
-	nbClient          *netboxdiodeplugin.Client
+	redisClient       RedisClient
+	redisStreamClient RedisClient
+	nbClient          netboxdiodeplugin.NetBoxAPI
```
Comment (mfiedorowicz) [inline diode-server/reconciler/ingestion_processor.go:65]: 👍🏻

File: diode-server/go.mod (hunk 1 above)
Comment (leoparente) [reply diode-server/go.mod]: done

File: diode-server/reconciler/ingestion_processor_internal_test.go (hunk 3 above)
Comment (leoparente) [reply diode-server/reconciler/ingestion_processor_internal_test.go]: done

File: diode-server/reconciler/ingestion_processor_internal_test.go (hunk 4 above)
Comment (leoparente) [reply diode-server/reconciler/ingestion_processor_internal_test.go]: Oh that is really cool. Just updated the PR using it! Great finding

File: diode-server/reconciler/ingestion_processor_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+	"github.com/netboxlabs/diode/diode-server/gen/diode/v1/diodepb"
+	"github.com/netboxlabs/diode/diode-server/netbox"
+	"github.com/netboxlabs/diode/diode-server/netboxdiodeplugin"
+	mnp "github.com/netboxlabs/diode/diode-server/netboxdiodeplugin/mocks"
+	"github.com/netboxlabs/diode/diode-server/reconciler/changeset"
+	mr "github.com/netboxlabs/diode/diode-server/reconciler/mocks"
+)
+
+func strPtr(s string) *string { return &s }
+func TestWriteJSON(t *testing.T) {
```
Comment (mfiedorowicz) [inline diode-server/reconciler/ingestion_processor_internal_test.go]: nit: can we add line break between these functions, please?

File: diode-server/reconciler/ingestion_processor_test.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+	"log/slog"
+
+	"github.com/alicebob/miniredis/v2"
+	"github.com/netboxlabs/diode/diode-server/gen/diode/v1/diodepb"
+	"github.com/netboxlabs/diode/diode-server/reconciler"
+	"github.com/redis/go-redis/v9"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/proto"
```
Comment (mfiedorowicz) [inline diode-server/reconciler/ingestion_processor_test.go]: goimport please

File: diode-server/reconciler/server_test.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	"testing"
+	"time"
+
+	"github.com/alicebob/miniredis/v2"
+	"github.com/netboxlabs/diode/diode-server/reconciler"
+	pb "github.com/netboxlabs/diode/diode-server/reconciler/v1/reconcilerpb"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
```
Comment (mfiedorowicz) [inline diode-server/reconciler/server_test.go]: goimport please

Comment (leoparente) [reply diode-server/reconciler/server_test.go]: done

File: diode-server/reconciler/ingestion_processor_test.go (hunk 7 above)
Comment (leoparente) [reply diode-server/reconciler/ingestion_processor_test.go]: done

File: diode-server/reconciler/ingestion_processor_internal_test.go (hunk 6 above)
Comment (leoparente) [reply diode-server/reconciler/ingestion_processor_internal_test.go]: done

**Review by mfiedorowicz — APPROVED** [review]
LGTM, thanks

**Pull request merged**